lib/sudo-cli/src/lib.rs

```diff
@@ -244,220 +75,268 @@ pub struct SudoOptions {
-            env_var_list: Default::default(),
-        })
-    }
+enum SudoArg {
+    Flag(String),
+    Argument(String, String),
+    Environment(String, String),
+    Rest(Vec<String>),
 }
 
 impl SudoOptions {
-    pub fn try_parse_from<I, T>(iter: I) -> Result<Self, Error>
+    const TAKES_ARGUMENT_SHORT: &[char] = &['C', 'D', 'E', 'g', 'h', 'p', 'R', 'T', 'U', 'u'];
+    const TAKES_ARGUMENT: &[&'static str] = &[
+        "close-from",
+        "chdir",
+        "preserve-env",
+        "group",
+        "host",
+        "chroot",
+        "command-timeout",
+        "other-user",
+        "user",
+    ];
+
+    /// argument assignments and shorthand options preprocessing
+    fn normalize_arguments<I>(iter: I) -> Result<Vec<SudoArg>, &'static str>
     where
-        I: IntoIterator<Item = T>,
-        T: Into<String> + Clone,
+        I: IntoIterator<Item = String>,
     {
-        // We need all this extra logic because `clap` cannot handle environment variable
-        // declarations. This means that we must filter any args with the shape `NAME=VALUE` before
-        // passing the arguments to `clap`.
-        //
-        // However, this is not straightforward because such args could be part of the command that
-        // `sudo` is supposed to execute. For example, in `sudo FOO=1 -b cmd BAR=2`, `FOO=1` should
-        // be filtered but `BAR=2` should not because it is part of the `cmd BAR=2` command.
-
-        // Keep the original arguments into `vec_args` in case we need them later.
-        let mut vec_args = iter.into_iter().map(Into::into).collect::<Vec<String>>();
-        let mut args = vec_args.iter();
-
-        // Store which args are environment variable declarations.
-        let mut vec_env_vars = vec![false; vec_args.len()];
-        let mut env_vars = vec_env_vars.iter_mut();
+        // the first argument is the sudo command - so we can sklip it
+        let mut arg_iter = iter.into_iter().skip(1).peekable();
+        let mut processed: Vec<SudoArg> = vec![];
 
-        // Store the arguments that were environment variable declarations.
-        let mut env_var_list = Vec::new();
-        // Store the arguments that were not environment variable declarations.
-        let mut remaining_args = Vec::new();
-        // Whether the args had a `"--"` to separate the external args from the regular args.
-        let mut had_separator = false;
+        while let Some(arg) = arg_iter.next() {
+            match arg.as_str() {
+                "--" => {
+                    processed.push(SudoArg::Rest(arg_iter.collect()));
+                    break;
+                }
+                long_arg if long_arg.starts_with("--") => {
+                    if long_arg.contains('=') {
+                        // convert assignment to normal tokens
+                        let (key, value) = long_arg.split_once('=').unwrap();
+                        processed.push(SudoArg::Argument(key.to_string(), value.to_string()));
```

Comment:
Should we not check for Self::TAKES_ARGUMENT here as well? Otherwise someone might do `--background=false` and think that this will turn that option off.